driver-core/src/main/com/mongodb/internal/connection/SocketChannelStream.java

```diff
@@ -50,11 +54,27 @@ public SocketChannelStream(final ServerAddress address, final SocketSettings set
         this.bufferProvider = notNull("bufferProvider", bufferProvider);
     }
 
+    private SocketChannel initializeSocketChannel() throws IOException {
+        Iterator<InetSocketAddress> inetSocketAddresses = address.getSocketAddresses().iterator();
+        while (inetSocketAddresses.hasNext()) {
+            SocketChannel socketChannel = SocketChannel.open();
+            try {
+                SocketStreamHelper.initialize(socketChannel.socket(), inetSocketAddresses.next(), settings, sslSettings);
+                return socketChannel;
+            } catch (SocketTimeoutException e) {
+                if (!inetSocketAddresses.hasNext()) {
+                    throw e;
```

Comment:
I thought it might confuse users to have a `SocketTimeoutException` that contains an alternative inetSocketAddress than the one they provided. However, seeing that it will be wrapped in a MongoSocketException in the open method alleviates that concern.

Could it be wrapped here instead? So that this method doesn't throw a IOException?

Reply:
This way keeps in consistent with SocketStream, which also needs to wrap IOException in its open method (because getInputStream and getOutputStream also throw that).  

I'm fine the way it is, but open to changing it if you feel strongly about it

Reply:
@jyemin good point. @thejonathanma lets leave this as is.